quickwit-janitor/src/actors/retention_policy_evaluator.rs

```diff
@@ -0,0 +1,460 @@
+
+    /// The number of evaluation passes.
+    pub num_evaluation_passes: usize,
+
+    /// The number of deleted splits.
+    pub num_discarded_splits: usize,
+}
+
+#[derive(Debug)]
+struct Loop;
+
+#[derive(Debug)]
+struct Evaluate {
+    index_id: String,
+}
+
+/// An actor for collecting garbage periodically from an index.
+pub struct RetentionPolicyEvaluator {
+    metastore: Arc<dyn Metastore>,
+    index_metadatas: HashMap<String, IndexMetadata>,
+    counters: RetentionPolicyEvaluatorCounters,
+}
+
+impl RetentionPolicyEvaluator {
+    pub fn new(metastore: Arc<dyn Metastore>) -> Self {
+        Self {
+            metastore,
+            index_metadatas: HashMap::new(),
+            counters: RetentionPolicyEvaluatorCounters::default(),
+        }
+    }
+
+    /// Refresh Loop handler logic.
+    /// Should not return an error to prevent the actor from crashing.
+    async fn handle_refresh_loop(&mut self, ctx: &ActorContext<Self>) {
+        info!("retention-policy-refresh-indexes-operation");
+        self.counters.num_refresh_passes += 1;
+
+        let index_metadatas = match self.metastore.list_indexes_metadatas().await {
+            Ok(metadatas) => metadatas,
+            Err(error) => {
+                error!(error=?error, "Failed to list indexes from the metastore.");
+                return;
+            }
+        };
+        info!(index_ids=%index_metadatas.iter().map(|im| &im.index_id).join(", "), "Retention policy refresh.");
+
+        let deleted_indexes = compute_deleted_indexes(
+            self.index_metadatas.keys(),
+            index_metadatas.iter().map(|metadata| &metadata.index_id),
+        );
+        for index_id in &deleted_indexes {
+            self.index_metadatas.remove(index_id);
+        }
+        info!(index_ids=%deleted_indexes.iter().join(", "), "Deleted indexes from cache.");
+
+        for index_metadata in index_metadatas.into_iter() {
+            ctx.record_progress();
+
+            // Only care about indexes with retention policy.
+            let retention_policy = match &index_metadata.retention_policy {
+                Some(policy) => policy,
+                None => {
+                    // Retention policy might have been removed.
+                    self.index_metadatas.remove(&index_metadata.index_id);
+                    continue;
+                }
+            };
+
+            // Insert or update index
+            if let Some(value) = self.index_metadatas.get_mut(&index_metadata.index_id) {
+                // Update cache value in case retention policy has changed.
+                *value = index_metadata;
+                continue;
+            }
+
+            let message = Evaluate {
+                index_id: index_metadata.index_id.clone(),
+            };
+            let next_interval = retention_policy
+                .duration_till_next_evaluation()
+                .expect("Expected a valid duration from retention policy schedule.");
```

Comment:
This is going to panic and crash the unsupervised actor. Log the error and skip the index.

Reply:
Yes related to the above comment. in principle we should always get a valid  duration_till_next_interval